internal/app/machined/pkg/controllers/k8s/apiserver_endpoints.go

```diff
@@ -0,0 +1,185 @@
+
+// Name implements controller.Controller interface.
+func (ctrl *APIServerEndpointsController) Name() string {
+	return "cluster.APIServerEndpointsController"
+}
+
+// Inputs implements controller.Controller interface.
+func (ctrl *APIServerEndpointsController) Inputs() []controller.Input {
+	return []controller.Input{
+		{
+			Namespace: config.NamespaceName,
+			Type:      config.MachineTypeType,
+			ID:        pointer.To(config.MachineTypeID),
+			Kind:      controller.InputWeak,
+		},
+		safe.Input[*cluster.Member](controller.InputWeak),
+		safe.Input[*config.MachineConfig](controller.InputWeak),
+		safe.Input[*cluster.Identity](controller.InputWeak),
+	}
+}
+
+// Outputs implements controller.Controller interface.
+func (ctrl *APIServerEndpointsController) Outputs() []controller.Output {
+	return []controller.Output{
+		{
+			Type: k8s.APIServerEndpointsType,
+			Kind: controller.OutputExclusive,
+		},
+	}
+}
+
+// Run implements controller.Controller interface.
+//
+//nolint:gocyclo,cyclop
+func (ctrl *APIServerEndpointsController) Run(ctx context.Context, r controller.Runtime, logger *zap.Logger) error {
+	for {
+		if _, ok := channel.RecvWithContext(ctx, r.EventCh()); !ok && ctx.Err() != nil {
+			return nil //nolint:nilerr
+		}
+
+		machineConfig, err := safe.ReaderGetByID[*config.MachineConfig](ctx, r, config.V1Alpha1ID)
+		if err != nil {
+			if !state.IsNotFoundError(err) {
+				return fmt.Errorf("error getting machine config: %w", err)
+			}
+
+			continue
+		}
+
+		machineType, err := safe.ReaderGetByID[*config.MachineType](ctx, r, config.MachineTypeID)
+		if err != nil {
+			if !state.IsNotFoundError(err) {
+				return fmt.Errorf("error getting machine type: %w", err)
+			}
+
+			continue
+		}
+
+		identity, err := safe.ReaderGetByID[*cluster.Identity](ctx, r, cluster.LocalIdentity)
+		if err != nil {
+			if !state.IsNotFoundError(err) {
+				return fmt.Errorf("error getting local identity: %w", err)
+			}
+
+			continue
+		}
+
+		members, err := safe.ReaderListAll[*cluster.Member](ctx, r)
+		if err != nil {
+			return fmt.Errorf("error listing affiliates: %w", err)
+		}
+
+		var endpoints []k8s.APIServerEndpoint
+
+		ce := machineConfig.Config().Cluster().Endpoint()
+		if ce != nil {
+			endpoints = append(endpoints, k8s.APIServerEndpoint{
+				Host: ce.Hostname(),
+				Port: toPort(ce.Port()),
+			})
+		}
+
+		if machineType.MachineType() == machine.TypeControlPlane {
+			endpoints = append(endpoints, k8s.APIServerEndpoint{
+				Host: "localhost",
+				Port: uint32(machineConfig.Config().Cluster().LocalAPIServerPort()),
+			})
+		}
+
+		for it := safe.IteratorFromList(members); it.Next(); {
+			memberSpec := it.Value().TypedSpec()
+
+			if identity.TypedSpec().NodeID == memberSpec.NodeID {
+				// skip adding ourselves to the list
+				continue
+			}
+
+			if len(memberSpec.Addresses) > 0 && memberSpec.ControlPlane != nil {
+				endpoints = append(endpoints, k8s.APIServerEndpoint{
+					Host: memberSpec.Addresses[0].String(),
```

Comment:
we should add every address here, not the first one. the loadbalancer will remove endpoitns which don't work